### DISCUSSION TITLE
Improve performance of recursive queries

### DISCUSSION THREAD

**Discussion by pankajp** [duckdb/duckdb#2766] [General]
I have a recursive query which recurses upto 200 depth (graph traversal). Is there a way to improve the performance of such queries? Maybe by adding an index or so? 

I just stumbled upon the paper https://arxiv.org/pdf/2108.10540.pdf "Making RDBMSs Efficient on Graph Workloads Through
Predefined Joins" which implements something with duckdb to improve performance of such queries. Are you aware of this and is there any plans for integrating something like that into duckdb?

**Comment by Alex-Monahan** [comment]
Hello and thanks for using DuckDB! 
Have you already experimented with DuckDB's ART indexes? Did that help at all?
[Indexes Docs](https://duckdb.org/docs/sql/indexes) 
I've also used DuckDB for graph traversal types of problems, so it's definitely a good use case for it!

**Reply by pankajp** [reply]
Thanks for the hint. Adding index for the slow joins reported in profile indeed helped speed it up a bit. I was wondering how I would create an index on the CTE but I guess that's not needed and adding index on the other tables being joined is sufficient.

**Reply by Alex-Monahan** [reply]
You can also set this pragma on the connection ahead of time to force index joins where possible!
```sql
-- Force index joins where applicable
PRAGMA force_index_join;
```

**Reply by pankajp** [reply]
Thanks, I'll read up what `force_index_join` does, because "Force index joins where applicable" isn't very clear to me for my limited understanding of databases. Would it do something like sqlite3 automatic indexes https://www.sqlite.org/optoverview.html#autoindex ?

**Reply by Alex-Monahan** [reply]
Good question! That setting is just controlling the optimizer to make sure that it always uses an index when it exists and is involved in a join. It won't automatically create indexes, but it just makes sure to use them when they exist. Does that help?
